src/coreclr/nativeaot/System.Private.CoreLib/src/System/Threading/ObjectHeader.cs

```diff
@@ -183,32 +186,302 @@ public static unsafe void SetSyncEntryIndex(int* pHeader, int syncIndex)
+                newBits |= syncIndex | BIT_SBLK_IS_HASH_OR_SYNCBLKINDEX;
             }
             while (Interlocked.CompareExchange(ref *pHeader, newBits, oldBits) != oldBits);
         }
+
+        //
+        // A few words about spinning choices:
+        //
+        // Most locks are not contentious. In fact most locks provide exclusive access safety, but in reality are used by
+        // one thread. And then there are locks that do see multiple threads, but the accesses are short and not overlapping.
+        // Thin lock is an optimization for such scenarios.
+        //
+        // If we see a thin lock held by other thread for longer than ~5 microseconds, we will "inflate" the lock
+        // and let the adaptive spinning in the fat Lock sort it out whether we have a contentious lock or a long-held lock.
+        //
+        // Another thing to consider is that SpinWait(1) is calibrated to about 35-50 nanoseconds.
+        // It can take much longer only if nop/pause takes much longer, which it should not, as that would be getting
+        // close to the RAM latency.
+        //
+        // Considering that taking and releasing the lock takes 2 CAS instructions + some overhead, we can estimate shortest
+        // time the lock can be held to be in hundreds of nanoseconds. Thus it is unlikely to see more than
+        // 8-10 threads contending for the lock without inflating it. Therefore we can expect to acquire a thin lock in
+        // under 16 tries.
+        //
+        // As for the backoff strategy we have two choices:
+        // Exponential back-off with a lmit:
+        //   0, 1, 2, 4, 8, 8, 8, 8, 8, 8, 8, . . . .
+        //
+        // Linear back-off
+        //   0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, . . . .
+        //
+        // In this case these strategies are close in terms of average and worst case latency, so we will prefer linear
+        // back-off as it favors micro-contention scenario, which we expect.
+        //
+
+        // Returs:
+        //   1 - success
+        //   0 - failed
+        //   syncIndex - retry with the Lock
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static unsafe int Acquire(object obj)
+        {
+            // Lock.s_processorCount is lazy-initialized at fist contention.
+            // untill then assume multicore
+            return TryAcquire(obj, oneShot: false);
+        }
+
+        // 1 - success
+        // 0 - failed
+        // syncIndex - retry with the Lock
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static unsafe int TryAcquire(object obj, bool oneShot = true)
+        {
+            ArgumentNullException.ThrowIfNull(obj);
+
+            Debug.Assert(!(obj is Lock),
+                "Do not use Monitor.Enter or TryEnter on a Lock instance; use Lock methods directly instead.");
+
+            int currentThreadID = Environment.CurrentManagedThreadIdUnchecked;
```

Comment:
I think it would look better for the unchecked getter to live on `ManagedThreadId` to have it closer to the dependency on implementation details.

Reply:
I was trying to avoid calling NativeAOT specific things from here - in case we share this with CoreCLR one day (I actually tried, it works, but a bit slower than native, for the same reasons).

Until then there is no need for this to be in the common location, and even then it may not be `Environment`, so, yes this can be in `ManagedThreadId`.

Reply:
And if when we switch to threadstatics, we can find a way to guarantee that ID is always initialized prior these uses, we will not need "Unchecked", because the regular accessor will not check either.

I will move this to `ManagedThreadId`

Reply:
> And if when we switch to threadstatics, we can find a way to guarantee that ID is always initialized prior these uses,

Where would that be?

Reply:
We initialize a bunch of stuff at time a thread enters managed runtime. Perhaps around there. I think that code even takes locks (in `Lock` sense), so may already indirectly initialize the ID, so we could make it a bit more intentional. 

I did not think much about this scenario though.

Reply:
`bunch of stuff` is not a lot in this case, we do try make this init minimal. but dispensing a thread ID is not very heavy.

Reply:
Right, we just add the thread to the ThreadStore when the thread enters managed runtime for the first time: https://github.com/dotnet/runtime/blob/7175e861011e323295f5fcfc0a5fac648974b80d/src/coreclr/nativeaot/Runtime/threadstore.cpp#L102-L142 . It is very little. There are no allocations or any managed code running during that operation. I do not think we would want to change that by adding thread ID dispensing.

Reply:
The was some code called from reverse pinvoke that runs threadstatic initialization and the like. I may be confusing this with module initialization.

Reply:
Yes, I was confusing things. The reverse pinvoke does access the Thread, but it accesses the _native_ Thread. It does not need managed ID and does not run managed code.  

There is also module init that runs managed code that executes static constructors and takes a Lock.

When thread is created from managed code the thread start method will set the managed thread ID. Entering via reverse pinvoke may leave the ID uninitialized though.

Reply:
BTW, the easiest way to hit "uninitialized ID" path is by using `lock(this)` in a finalizer. Finalizer thread reverse-pinvokes into finalizer methods.